src/core/exec-invoke.c

```diff
@@ -2563,12 +2617,41 @@ static int setup_exec_directory(
                 if (params->runtime_scope != RUNTIME_SCOPE_SYSTEM)
                         continue;
 
-                /* Then, change the ownership of the whole tree, if necessary. When dynamic users are used we
+                idmapping_supported = is_idmapping_supported(pp ?: p);
+                if (idmapping_supported < 0)
+                        goto fail;
+
+                log_debug("ID-mapping is%ssupported for exec directory %s", idmapping_supported ? " " : " not ", pp ?: p);
+
+                /* Change the ownership of the whole tree, if necessary. When dynamic users are used we
                  * drop the suid/sgid bits, since we really don't want SUID/SGID files for dynamic UID/GID
                  * assignments to exist. */
-                r = path_chown_recursive(pp ?: p, uid, gid, context->dynamic_user ? 01777 : 07777, AT_SYMLINK_FOLLOW);
-                if (r < 0)
-                        goto fail;
+                if (uid == 0 || gid == 0 || !idmapping_supported)
+                        do_chown = true;
```

Comment:
maybe set `i->idmapped = false` here. Not because this was necessary, but more as code-enforced documentation